pkg/storage/bloom/bloom-shipper/bloom_shipper.go

```diff
@@ -0,0 +1,296 @@
+package bloom_shipper
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/grafana/loki/pkg/storage"
+	"github.com/grafana/loki/pkg/storage/chunk/client"
+	"github.com/grafana/loki/pkg/storage/config"
+	"github.com/grafana/loki/pkg/util/math"
+	"github.com/prometheus/common/model"
+	"io"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const (
+	metasFolder           = "metas"
+	delimiter             = "/"
+	fileNamePartDelimiter = "-"
+)
+
+type BlockRef struct {
+	TenantID      string
+	TSDBSource    string
+	BlockFilePath string
+	//todo check if it should be string
+	Checksum string
+
+	MinFingerprint, MaxFingerprint uint64
+	StartTimestamp, EndTimestamp   int64
+}
+
+type MetaRef struct {
+	TenantID                       string
+	TableName                      string
+	MinFingerprint, MaxFingerprint uint64
+	// check if it has to be time.Time
+	StartTimestamp, EndTimestamp int64
+	//todo check if it should be string
+	Checksum string
+	FilePath string
+}
+
+// todo rename it
+type Meta struct {
+	MetaRef `json:"-"`
+
+	Tombstones []BlockRef
+	Blocks     []BlockRef
+}
+
+type MetaSearchParams struct {
+	TenantID       string
+	MinFingerprint uint64
+	MaxFingerprint uint64
+	StartTimestamp int64
+	EndTimestamp   int64
+}
+
+type MetaShipper interface {
+	// Returns all metas that are within MinFingerprint-MaxFingerprint fingerprint range
+	// and intersect time period from StartTimestamp to EndTimestamp.
+	GetAll(ctx context.Context, params MetaSearchParams) ([]Meta, error)
+	Upload(ctx context.Context, meta Meta) error
+	Delete(ctx context.Context, meta Meta) error
+}
+
+type Block struct {
+	BlockRef
+	//todo TBD
+	Data []byte
+}
+
+type BlockShipper interface {
+	GetBlocks(ctx context.Context, references []BlockRef) ([]Block, error)
+	UploadBlocks(ctx context.Context, blocks []Block) error
+	DeleteBlocks(ctx context.Context, blocks []BlockRef) error
+}
+
+type Shipper interface {
+	MetaShipper
+	BlockShipper
+	Stop()
+}
+
+// todo add logger
+func NewShipper(periodicConfigs []config.PeriodConfig, storageConfig storage.Config, clientMetrics storage.ClientMetrics) (*bloomShipper, error) {
+	periodicObjectClients := make(map[config.DayTime]client.ObjectClient)
+	for _, periodicConfig := range periodicConfigs {
+		objectClient, err := storage.NewObjectClient(periodicConfig.ObjectType, storageConfig, clientMetrics)
```

Comment:
must be prefixedObjectClient . see: tsdb.shared-store-key-prefix

Reply:
currently, it's not possible to use `prefixedObjectClient` outside of package `storage`.